proposals/0000-custom-type-errors.rst

```diff
@@ -0,0 +1,188 @@
+
+
+Costs and Drawbacks
+-------------------
+At first glance, the development and maintenance costs should be small.
+
+In principle, this extension should make libraries easier to learn. One problem remains, though: how to show type signatures in an understandable way. For most people ::
+
+    (==) :: Eq a => a -> a -> Bool
+
+conveys much more meaning that the long signature. Whether we should show one or the other in the interactive environment is discussed as an unresolved questions.
+
+One drawback of this approach is that it requires library authors to use ``DataKinds`` whenever they want to use custom errors. As a side-effect, some of the type definitions in a module might be promoted to the kind level. Right now we do not have a way to control data type promotion, so this is unavoidable.
+
+
+Alternatives
+------------
+In this proposal, ``IfUndischarged`` is associated to a constraint in each usage site. Another alternative is having the message attached to the type class itself. This is the route taken by Scala with their `implicitNotFound annotation <http://www.scala-lang.org/api/2.12.0/scala/annotation/implicitNotFound.html>`_
+
+That alternative, however, is less flexible than the current proposal, since you could always export a new constraint which includes the message ::
+
+    type Eq' a = IfUndischarged (Eq a) (Text "blah blah")
+
+Note, however, that this approach has the drawback of having different names for the annotated (``Eq'``) and original (``Eq``) type classes. The user has to remember that the latter should be used when writing a new instance, but the former when writing the signature of a function.
+
+
+Unresolved questions
+--------------------
+
+Messages or hints?
+~~~~~~~~~~~~~~~~~~
+In the description above, I have replaced the default error messages by custom ones completely. Maybe a better choice, especially for ``IfUndischarghed``, it to add the information as a *hint* or *suggestion*, in addition to the default message.
+
+How to print the annotated signature?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In the current prototype, as a side-effect of the way in which GHCi computes the type to print, the signature of a function is always simplified. In this case, that means that no trace of ``IfUndischarged`` or ``IfApart`` is shown. Is this the right behavior? (I think it is)
+
+The same question should be asked about Haddock. Maybe the smallest, simplified signature should be the one in the main documentation, and the error information should get some specific markup. Of course, this means that now Haddock has to inspect the types of the documented values, something which is not done as of now.
+
+Generalization
+~~~~~~~~~~~~~~
+What happens if we need to infer a type with a constraint which has an attached message? Do we at it using ``IfUndischarged``? This definitely seems like a wrong path, although it is also surprising that if I write ::
```

Comment:
This is an interesting point, but this problem exists today. There is no guarantee that `x = y` will infer a type for `x` that equals the type for `y`. For example, if `y`'s type is `(Eq a, Ord a) => a -> a`, then `x`'s will just mention the `Ord` constraint.